### PR TITLE
Add logs for nul content corruption issue

### PR DESCRIPTION
**Pull request by donho**
These logs are written for debugging NUL file-corruption issue (#6133) and session lost issue (#10402) which are still problematic probably.

To activate log, user should:
1. Add an empty "nppLogNulContentCorruptionIssue.xml" file beside of notepad++.exe, or if user has no admin privilege, he/she can add this file into %APPDATA%\Notepad++\.
2. Start notepad++.exe, all saving traces will be in the file named "nppLogNulContentCorruptionIssue.log" in %APPDATA%\Notepad++\, or in Notepad++ installation dir if doLocalConf.xml is present.

### CODE REVIEW

**Comment by donho** [comment]
@pnedev 
What kind of logs are interesting to add besides the current ones in this PR?

**Comment by pnedev** [comment]
@donho ,
The main reasons behind changing the file API were the NUL problem and the automatic session loss and both of these seem to appear only when Windows restarts for applying automatic updates. Because this is not easily reproducible I was thinking if we could try to create a 'hidden' (from the user) log file when Notepad++ is notified about the coming restart. It will log if session was saved before the restart and if there are unsaved files or whatever you might think of. I know that the log itself may be corrupted and invalid in that event but it might as well succeed and next time a user hits this problem we'll ask him to provide this log to see what happened in that auto-restart situation - was Notepad++ notified, where any errors reported when saving session, etc.

I haven't looked yet at the Notepad++ code dealing with system restart notification but the problem is appearing on that event and it will be interesting to log what is happening at that time (if log succeeds at all but lets try).

Nothing else comes to my mind now.

**Comment by donho** [comment]
> Because this is not easily reproducible I was thinking if we could try to create a 'hidden' (from the user) log file when Notepad++ is notified about the coming restart. It will log if session was saved before the restart and if there are unsaved files or whatever you might think of. 

Good thinking! I'll try to do in this direction.

> I haven't looked yet at the Notepad++ code dealing with system restart notification but the problem is appearing on that event and it will be interesting to log what is happening at that time (if log succeeds at all but lets try).

Here's the log complete of 1 session with Windows 11 update - restart manually while Notepad++ is opened:

```
2021-12-02 23:18:28  WM_CREATE ===========================================
2021-12-02 23:18:35  C:\Users\userX\AppData\Roaming\Notepad++\backup\test.txt@2021-12-02_231835 is opened.
2021-12-02 23:18:35  C:\Users\userX\AppData\Roaming\Notepad++\backup\test.txt@2021-12-02_231835  202/202 bytes are written.
2021-12-02 23:18:35  C:\Users\userX\AppData\Roaming\Notepad++\backup\test.txt@2021-12-02_231835 is closed.
2021-12-02 23:18:35  C:\Users\userX\AppData\Roaming\Notepad++\session.xml is opened.
2021-12-02 23:18:35  C:\Users\userX\AppData\Roaming\Notepad++\session.xml  1377/1377 bytes are written.
2021-12-02 23:18:35  C:\Users\userX\AppData\Roaming\Notepad++\session.xml is closed.
2021-12-02 23:18:46  C:\aDirectory\test.txt is opened.
2021-12-02 23:18:46  C:\aDirectory\test.txt  202/202 bytes are written.
2021-12-02 23:18:46  C:\aDirectory\test.txt is closed.
2021-12-02 23:19:03  C:\Users\userX\AppData\Roaming\Notepad++\backup\test.txt@2021-12-02_231903 is opened.
2021-12-02 23:19:03  C:\Users\userX\AppData\Roaming\Notepad++\backup\test.txt@2021-12-02_231903  578/578 bytes are written.
2021-12-02 23:19:03  C:\Users\userX\AppData\Roaming\Notepad++\backup\test.txt@2021-12-02_231903 is closed.
2021-12-02 23:19:03  C:\Users\userX\AppData\Roaming\Notepad++\session.xml is opened.
2021-12-02 23:19:03  C:\Users\userX\AppData\Roaming\Notepad++\session.xml  1379/1379 bytes are written.
2021-12-02 23:19:03  C:\Users\userX\AppData\Roaming\Notepad++\session.xml is closed.
2021-12-02 23:19:26  WM_QUERYENDSESSION
2021-12-02 23:19:26  C:\Users\userX\AppData\Roaming\Notepad++\config.xml is opened.
2021-12-02 23:19:26  C:\Users\userX\AppData\Roaming\Notepad++\config.xml  7230/7230 bytes are written.
2021-12-02 23:19:26  C:\Users\userX\AppData\Roaming\Notepad++\config.xml is closed.
2021-12-02 23:19:26  C:\Users\userX\AppData\Roaming\Notepad++\session.xml is opened.
2021-12-02 23:19:26  C:\Users\userX\AppData\Roaming\Notepad++\session.xml  1379/1379 bytes are written.
2021-12-02 23:19:26  C:\Users\userX\AppData\Roaming\Notepad++\session.xml is closed.
2021-12-02 23:19:26  WM_ENDSESSION
2021-12-02 23:19:26  WM_DESTROY
```

Notepad++ process WM_QUERYENDSESSION message, in which you can check.

File: PowerEditor/src/MISC/Common/FileInterface.cpp [hunk 1]
```diff
@@ -33,40 +33,48 @@ Win32_IO_File::Win32_IO_File(const wchar_t *fname)
...
-		_path = std::string(fn.begin(), fn.end());
+		std::wstring fn = fname;
+		std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> converter;
+		_path = converter.to_bytes(fn);
 		_hFile = ::CreateFileW(fname, _accessParam, _shareParam, NULL, _dispParam, _attribParam, NULL);
+
+
+		if (NppParameters::getInstance().doNppLogNulContentCorruptionIssue())
+		{
+			generic_string nppLogNetworkDriveIssueLog = TEXT("c:\\temp\\");
```
Comment (mere-human) [inline PowerEditor/src/MISC/Common/FileInterface.cpp:44]: Wouldn't it be easier just to use %UserProfile% that always exists? So the user is not obligated to create it beforehand.

Comment (donho) [reply PowerEditor/src/MISC/Common/FileInterface.cpp]: @mere-human 
Yes, indeed.
In the newest commit `%APPDATA%\Notepad++\` is used for storing the log files.

**Comment by donho** [comment]
@pnedev 
Just did the last commit by following your suggestion:
Enabled by default (nppLogNulContentCorruptionIssue.xml will be included in all packages), and the log will be triggered only when Windows notifies Notepad++ to quit. I just triggered another update restarting manually, and here are the logs:
```
2021-12-03 20:31:53  WM_QUERYENDSESSION =====================================
2021-12-03 20:31:53  C:\Users\donh\AppData\Roaming\Notepad++\config.xml is opened.
2021-12-03 20:31:53  C:\Users\donh\AppData\Roaming\Notepad++\config.xml  7754/7754 bytes are written.
2021-12-03 20:31:53  C:\Users\donh\AppData\Roaming\Notepad++\config.xml is closed.
2021-12-03 20:31:53  C:\Users\donh\AppData\Roaming\Notepad++\session.xml is opened.
2021-12-03 20:31:53  C:\Users\donh\AppData\Roaming\Notepad++\session.xml  1969/1969 bytes are written.
2021-12-03 20:31:53  C:\Users\donh\AppData\Roaming\Notepad++\session.xml is closed.
2021-12-03 20:31:53  WM_ENDSESSION
2021-12-03 20:31:53  WM_DESTROY
```
Sorry for the unrelated modifications - VS2022 doesn't take override function for certain (unknown for me) conditions.

**Comment by xomx** [comment]
@donho 
> and the log will be triggered only when Windows notifies Notepad++ to quit

I tried the 8.1.9.3RC (x64 portable) and my log is filled only by `WM_DESTROY` lines for every N++ session closed.
Is that desired behavior?

**Comment by donho** [comment]
@xomx 
It's not desired behaviour - if Notepad++ is closed normally, there won't be any log written. 
The logs are written only when Windows is about to restart and Notepad++ is still open.
The bug has been fixed in https://github.com/notepad-plus-plus/notepad-plus-plus/commit/5d2fc0d8c08e44afec176b3c37987810897e8a5c
I'll do another RC - thank you for your head up!

**Pull request closed without merging**